pkg/exemplars/tsdb.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package exemplars
+
+import (
+	"github.com/gogo/status"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/storage"
+	"google.golang.org/grpc/codes"
+
+	"github.com/thanos-io/thanos/pkg/exemplars/exemplarspb"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+)
+
+// TSDB implements exemplarspb.Exemplars that allows to fetch exemplars from a TSDB instance.
+type TSDB struct {
+	db        storage.ExemplarQueryable
+	extLabels labels.Labels
+}
+
+// NewTSDB creates new exemplars.TSDB.
+func NewTSDB(db storage.ExemplarQueryable, extLabels labels.Labels) *TSDB {
+	return &TSDB{
+		db:        db,
+		extLabels: extLabels,
+	}
+}
+
+// Exemplars returns all specified exemplars from a TSDB instance.
+func (t *TSDB) Exemplars(r *exemplarspb.ExemplarsRequest, s exemplarspb.Exemplars_ExemplarsServer) error {
+	expr, err := parser.ParseExpr(r.Query)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	match, selectors := selectorsMatchesExternalLabels(parser.ExtractSelectors(expr), t.extLabels)
```

Comment:
Again the query parsing is still duplicated. Why not just use the parsed matchers in the method signature?

``` go
func (t *TSDB) Exemplars(matchers [][]*labels.Matcher, s exemplarspb.Exemplars_ExemplarsServer) error { ... }
```

Reply:
Now I got what you meant, we are doing query parsing for each tenant, yeah that's not good.

But we can't change the signature I guess, If we change the signature, `*TSDB` will not implement `exemplarspb.ExemplarsServer` interface.

Reply:
One option is to put `[][]*label.Matchers` (or labelpb equivalent) in the `ExemplarsRequest` instead of the raw query. For sidecars, maybe we can reconstruct the query using the matches but I am not sure about this part.

Reply:
I am wondering why we need to implement this interface for `TSDB`. It is already implemented in `MultiTSDB`, right? For each tenant's TSDB, as it is part of the multitsdb. Can we access it by a separate function?

Reply:
Actually, I don't think we have any hard requirement to implement `exemplarspb.ExemplarsServer` for `exemplars.TSDB`, let's do what you suggested.

Reply:
Yes, that would be better. We can change the part https://github.com/thanos-io/thanos/pull/4292/files#diff-970995d3ac812ca4b94e73daf4773bd1a3f80cd746ef697e2084a29a4802333aR27-R34 and use a new method for TSDB exemplars.

Reply:
Done, pushed a commit with this (https://github.com/thanos-io/thanos/pull/4292/commits/f6a00371c80784d9a479120b56cd8409bddc40f8)